### ISSUE TITLE
API Post with missing hostname updates all hosts

### ISSUE THREAD

**Issue by mzac** [Icinga/icinga2#5407]
One of my database scripts was doing an API POST to update specific hosts in Icinga2, however someone left the hostname field blank in the database.  When the post is run with no hostname, all host objects in Icinga2 get updated with the IP of the blank entry.

## Expected Behavior

Icinga2 should not accept a blank hostname when doing a 'POST'

## Current Behavior

All hosts get their IP replaced with the wrong IP address, including hosts that were added with .conf files.

## Possible Solution

Find a way for the API to reject POSTS that don't contain only a host template and a hostname

## Steps to Reproduce (for bugs)

1. Create a bunch of hosts
2. Send a API Post with the host template and ip address and leave the hostname empty

Test adding hosts via API:

```
curl -k -s -u 'icinga2-director:jDkYd3bKkMLs4MK' \
-H 'Accept: application/json' \
-X PUT 'https://localhost:5665/v1/objects/hosts/test10' \
-d '{"templates": [ "generic-host" ], "attrs": {"address": "127.0.0.10" }}' \
| python -m json.tool 

curl -k -s -u 'icinga2-director:jDkYd3bKkMLs4MK' \
-H 'Accept: application/json' \
-X PUT 'https://localhost:5665/v1/objects/hosts/test11' \
-d '{"templates": [ "generic-host" ], "attrs": {"address": "127.0.0.11" }}' \
| python -m json.tool 

curl -k -s -u 'icinga2-director:jDkYd3bKkMLs4MK' \
-H 'Accept: application/json' \
-X PUT 'https://localhost:5665/v1/objects/hosts/test12' \
-d '{"templates": [ "generic-host" ], "attrs": {"address": "127.0.0.12" }}' \
| python -m json.tool 
```

Submit a 'POST' with blank hostname and a new IP address, all hosts get this new IP address:

```
root@a5042a459b4e:/etc/icinga2/conf.d# curl -k -s -u 'icinga2-director:jDkYd3bKkMLs4MK' \
> -H 'Accept: application/json' \
> -X POST 'https://localhost:5665/v1/objects/hosts/' \
> -d '{"templates": [ "generic-host" ], "attrs": {"address": "127.0.0.22" }}' \
> | python -m json.tool
{
    "results": [
        {
            "code": 200.0,
            "name": "a5042a459b4e",
            "status": "Attributes updated.",
            "type": "Host"
        },
        {
            "code": 200.0,
            "name": "test10",
            "status": "Attributes updated.",
            "type": "Host"
        },
        {
            "code": 200.0,
            "name": "test11",
            "status": "Attributes updated.",
            "type": "Host"
        },
        {
            "code": 200.0,
            "name": "test12",
            "status": "Attributes updated.",
            "type": "Host"
        }
    ]
}
```

## Context

It basically wiped out all my monitoring.  I had to stop Icinga2 completely, delete all my state files and start from scratch.

## Your Environment

* Version used (`icinga2 --version`): 2.6.3
* Operating System and version: Redhat 7.3
* Enabled features (`icinga2 feature list`): api checker command ido-mysql mainlog notification statusdata
* Icinga Web 2 version and modules (System - About): 2.4.1
* Config validation (`icinga2 daemon -C`): 
* If you run multiple Icinga 2 instances, the `zones.conf` file (or `icinga2 object list --type Endpoint` and `icinga2 object list --type Zone`) from all affected nodes.

**Comment by Crunsher** [comment]
Updating everything matching your mask is by design, so it's not a bug per se...
But this can lead to problems like this. @gunnarbeutner @dnsmichi what is your take on this?

**Comment by mzac** [comment]
Yes, I kind of figured it was as you say 'by design'... I fixed my database scripts to my SQL queries would make sure there was a name when selecting from my database, but that still doesn't fix the fact that a whole Icinga config can be overwritten with one command!

**Comment by Stefar77** [comment]
@Crunsher Maybe update multiple hosts only when you add a "multiple": true tag in the JSON or only when hostname equals "*" or something?

**Comment by Crunsher** [comment]
@Stefar77 This would require us the API a version and not solve the problem for most users :(

**Comment by mzac** [comment]
@Crunsher I agree that probably a new v2 would have to be made to do this kind of fix.

Also, something else I was wondering with this situation.  In my setup I have about 40% of my hosts that are created in .conf files and the rest through the API.  When changing the IP of hosts created with .conf files via the API, how does one get the original IP back set in the .conf file?

Is what I did by deleting the state files etc the only fix?  Can I force Icinga to re-read the .conf files any other way?

**Comment by dnsmichi** [comment]
`/hosts/` could somehow being detected, and an error being thrown if no `filter` is set. But I don't think that this will entirely solve the issue, nor raise awareness on "defensive programming" with the API. I would *always* check that my attributes are safe and sound before pushing something harmful to an external application. 

The initial post makes me wonder - are you testing in production? I highly recommend building and testing your application in a dev instance beforehand. If you need Icinga 2 code to generate lots of hosts and services, you might look into the existing Vagrant box environments.

**Comment by mzac** [comment]
@dnsmichi I wasn't testing in production... it 'happened' on my production system so then I tested in a docker image.  I have since fixed my database scripts so it 'should not happen again'.

And yes for sure, I agree that checks could always be made, but people make mistakes and especially since more and more applications are using APIs now, we are starting to integrate with more applications in all directions.  I can make all the checks I want, but if someone else pushes to my Icinga2 API and makes a mistake, well there isn't anything I can do about that but get upset with them :)

So all that being said, yes, maybe a check for this kind of situation where no hostname is specified and only the IP address field is trying to be updated would be a good check.  Also, there is the cascade option that is used?  Maybe it could be used in this situation if needed instead of the 'multiple' or * ?
